src/lib/core/internal/internals.ts

```diff
@@ -0,0 +1,71 @@
+import { Observable } from 'rxjs';
+
+/**
+ * @property type - Dispatched event type
+ * @property payload - Paylaod to dispatch
+ */
+export interface DispatchedEvent<T = unknown> {
+    type: string;
+    payload?: T;
+}
+
+/**
+ * Methods decorated with `@Dispatch()` decorator can return plain object or `Promise` or `Observable`
+ */
+export type WrappedDispatchedEvent = Observable<DispatchedEvent> | Promise<DispatchedEvent> | DispatchedEvent;
+
+/**
+ * Event can be a plain object or an instance of some class
+ *
+ * @internal
+ * @param event - Dispatched event
+ * @returns - True if plain object or instance constructor has `type` property
+ */
+export function hasTypeProperty<T extends Object>(event: T): boolean {
+    return event.hasOwnProperty('type') || event.constructor.hasOwnProperty('type');
+}
+
+/**
+ * @internal
+ * @param event - Dispatched event
+ * @returns - True if event is an object and has `type` property
+ */
+export function isValidEvent<T extends Object>(event: T): boolean {
+    return !!event && typeof event === 'object' && hasTypeProperty<T>(event);
+}
+
+/**
+ * Descriptor exists only in case of method decorating
+ *
+ * @internal
+ * @param descriptor - Property descriptor
+ * @returns - True if descriptor exists
+ */
+export function descriptorExists(descriptor: TypedPropertyDescriptor<Function> | undefined): boolean {
+    return !!descriptor && descriptor.hasOwnProperty('value');
+}
+
+/**
+ * @internal
+ * @param event - Dispatched event
+ * @returns - True if event is just a plain object without constructor
+ */
+export function eventIsPlainObject<T extends Object>(event: T): boolean {
+    return event.constructor === Object;
```

Comment:
```ts
Object(event) === event
```

Reply:
This is interesting. Is there a cannonical source for these checks that we can reference in our source code?

Reply:
![image](https://user-images.githubusercontent.com/12021443/48998392-6ba81000-f164-11e8-8d2a-1a1a67d2baf0.png)

Reply:
@splincode That changes the behavior. I think that the check is for a plain old javascrip object.
This is a more robust implementation:
![image](https://user-images.githubusercontent.com/1948265/48999675-094d1080-f160-11e8-8ed2-a678ca399c84.png)

Reply:
The problem is - I'm not checking if object's constructor is some class. I want only `isPlainObject({})` return `true` and in other cases like `isPlainObject(new A())` - should return false, as its constructor is `class A`

Reply:
To me the `isPlainObject` name makes sense (describes the 'what'). `isObjectWithoutConstructor` describes the 'how' which is not what we would want.
There is such as concept as a POJO : Plain Old Javascript Object
Maybe a more recognisable term is an 'Object Literal'.
Could call it `isObjectLiteral`